File: crabzilla-vavr-example1/crabzilla-vavr-example1-service/src/test/java/io/github/crabzilla/example1/Example1AcceptanceIT.java
package io.github.crabzilla.example1;

import com.jayway.restassured.RestAssured;
import com.jayway.restassured.http.ContentType;
import io.github.crabzilla.core.entity.EntityUnitOfWork;
import io.github.crabzilla.core.entity.Version;
import io.github.crabzilla.example1.customer.Customer;
import io.vertx.core.json.Json;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.UUID;

import static com.jayway.restassured.RestAssured.given;
import static com.jayway.restassured.http.ContentType.JSON;
import static io.github.crabzilla.example1.customer.CustomerData.*;
import static io.github.crabzilla.vertx.helpers.StringHelper.aggregateId;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;

@Slf4j
public class Example1AcceptanceIT {

  static final String LOCATION_HEADER = "Location";

  @BeforeClass
  public static void configureRestAssured() throws InterruptedException {
    RestAssured.baseURI = "http://localhost";
    RestAssured.port = Integer.getInteger("http.port");
    log.info("----> RestAssured.port=" + RestAssured.port);
  }

  @AfterClass
  public static void unconfigureRestAssured() {
    RestAssured.reset();
  }

  // tag::create_customer_test[]

  @Test
  public void createCustomer() {

    val customerId = new CustomerId(UUID.randomUUID().toString());
    val createCustomerCmd = new CreateCustomer(UUID.randomUUID(), customerId, "customer test");
    val expectedEvent = new CustomerCreated(createCustomerCmd.getTargetId(), "customer test");
    val expectedUow = new EntityUnitOfWork(UUID.randomUUID(), createCustomerCmd,
            new Version(1), singletonList(expectedEvent));

    val json = Json.encodePrettily(createCustomerCmd);

    log.info("command=\n" + json);

    val postCmdResponse = given().
            contentType(JSON).
            body(json).
            when().
            post("/" + aggregateId(Customer.class) + "/commands").
            then().
            extract().response();

    assertThat(postCmdResponse.statusCode()).isEqualTo(201);
    assertThat(postCmdResponse.header(LOCATION_HEADER))
            .isEqualTo(RestAssured.baseURI + ":" + RestAssured.port + "/"
                    + aggregateId(Customer.class) + "/commands/" + createCustomerCmd.getCommandId().toString());

    val getUowResponse = given().
            contentType(JSON).
            body(json).
            when().
            get(postCmdResponse.header(LOCATION_HEADER)).
            then().
            statusCode(200).
            contentType(ContentType.JSON)
            .extract().response().asString();

    val uow = Json.decodeValue(getUowResponse, EntityUnitOfWork.class);

    assertThat(uow.targetId()).isEqualTo(expectedUow.targetId());
    assertThat(uow.getCommand()).isEqualTo(expectedUow.getCommand());
    assertThat(uow.getEvents()).isEqualTo(expectedUow.getEvents());
    assertThat(uow.getVersion()).isEqualTo(expectedUow.getVersion());

  }

  // end::create_customer_test[]


  @Test
  public void createCustomerIdempotency() {

    val customerId = new CustomerId(UUID.randomUUID().toString());
    val createCustomerCmd = new CreateCustomer(UUID.randomUUID(), customerId, "customer test");
    val expectedEvent = new CustomerCreated(createCustomerCmd.getTargetId(), "customer test");
    val expectedUow = new EntityUnitOfWork(UUID.randomUUID(), createCustomerCmd,
            new Version(1), singletonList(expectedEvent));

    val json = Json.encodePrettily(createCustomerCmd);

    log.info("command=\n" + json);

    val postCmdResponse = given().
            contentType(JSON).
            body(json).
            when().
            post("/" + aggregateId(Customer.class) + "/commands").
            then().
            extract().response();

    assertThat(postCmdResponse.statusCode()).isEqualTo(201);
    assertThat(postCmdResponse.header(LOCATION_HEADER))
            .isEqualTo(RestAssured.baseURI + ":" + RestAssured.port + "/"
                    + aggregateId(Customer.class) + "/commands/" + createCustomerCmd.getCommandId().toString());

    val getUowResponse = given().
            contentType(JSON).
            body(json).
            when().
            get(postCmdResponse.header(LOCATION_HEADER)).
            then().
            statusCode(200).
            contentType(ContentType.JSON)
            .extract().response().asString();

    val uow = Json.decodeValue(getUowResponse, EntityUnitOfWork.class);

    assertThat(uow.targetId()).isEqualTo(expectedUow.targetId());
    assertThat(uow.getCommand()).isEqualTo(expectedUow.getCommand());
    assertThat(uow.getEvents()).isEqualTo(expectedUow.getEvents());
    assertThat(uow.getVersion()).isEqualTo(expectedUow.getVersion());

    // now lets post it again

    val postCmdResponse2 = given().
            contentType(JSON).
            body(json).
            when().
            post("/" + aggregateId(Customer.class) + "/commands").
            then().
            extract().response();

    assertThat(postCmdResponse2.statusCode()).isEqualTo(201);
    assertThat(postCmdResponse2.header(LOCATION_HEADER))
            .isEqualTo(RestAssured.baseURI + ":" + RestAssured.port + "/"
                    + aggregateId(Customer.class) + "/commands/" + createCustomerCmd.getCommandId().toString());

  }

  @Test
  public void unknownCustomer() {

    val customerId = new CustomerId(UUID.randomUUID().toString());
    val activateCustomer = new ActivateCustomer(UUID.randomUUID(), customerId, "customer test");
    val json = Json.encodePrettily(activateCustomer);

    log.info("command=\n" + json);

    val response = given().
            contentType(JSON).
            body(json).
            when().
            post("/" + aggregateId(Customer.class) + "/commands").
            then().
            statusCode(400).
            extract().response().asString();

    assertThat(response).isEmpty();

  }

}

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-service/src/test/java/io/github/crabzilla/example1/Example1AcceptanceIT.kt
package io.github.crabzilla.example1

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule
import com.jayway.restassured.RestAssured
import com.jayway.restassured.RestAssured.given
import com.jayway.restassured.http.ContentType
import com.jayway.restassured.http.ContentType.JSON
import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.core.entity.EntityUnitOfWork
import io.github.crabzilla.core.entity.Version
import io.github.crabzilla.example1.customer.*
import io.github.crabzilla.vertx.helpers.StringHelper.aggregateId
import io.vertx.core.json.Json
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.*

class Example1AcceptanceIT {

  internal val LOCATION_HEADER = "Location"

  internal var mapper: ObjectMapper = Json.prettyMapper

  @Before
  @Throws(InterruptedException::class)
  fun configureRestAssured() {
    RestAssured.baseURI = "http://localhost"
    RestAssured.port = Integer.getInteger("http.port", 8080)!!
//    RestAssured.port = 8080

    log.info("----> RestAssured.port=" + RestAssured.port)
    mapper.registerModule(ParameterNamesModule())
            .registerModule(Jdk8Module())
            .registerModule(JavaTimeModule())
            .registerModule(KotlinModule())
            .enable(SerializationFeature.INDENT_OUTPUT)
  }

  @After
  fun unconfigureRestAssured() {
    RestAssured.reset()
  }

  // tag::create_customer_test[]

  @Test
  @Throws(IOException::class)
  fun createCustomer() {

    val customerId = CustomerId(UUID.randomUUID().toString())
    val createCustomerCmd = CreateCustomer(UUID.randomUUID(), customerId, "customer test")
    val expectedEvent = CustomerCreated(createCustomerCmd._targetId, "customer test")
    val expectedUow = EntityUnitOfWork(UUID.randomUUID(), createCustomerCmd,
            Version(1), listOf<DomainEvent>(expectedEvent))

    val json = mapper.writerFor(CreateCustomer::class.java).writeValueAsString(createCustomerCmd)

    log.info("command=\n" + json)

    val postCmdResponse = given().contentType(JSON).body(json)
            .`when`().post("/" + aggregateId(Customer::class.java) + "/commands")
            .then().extract().response()

    assertThat(postCmdResponse.statusCode()).isEqualTo(201)
    assertThat(postCmdResponse.header(LOCATION_HEADER))
            .isEqualTo(RestAssured.baseURI + ":" + RestAssured.port + "/"
                    + aggregateId(Customer::class.java) + "/commands/" + createCustomerCmd.commandId.toString())

    val getUowResponse = given().contentType(JSON).body(json)
            .`when`().get(postCmdResponse.header(LOCATION_HEADER))
            .then().statusCode(200).contentType(ContentType.JSON)
            .extract().response()

//    log.info("response -------------------")
//    log.info(getUowResponse.asString())

    val uow = mapper.readValue(getUowResponse.asString(), EntityUnitOfWork::class.java)

    assertThat(uow.targetId()).isEqualTo(expectedUow.targetId())
    assertThat(uow.command).isEqualTo(expectedUow.command)
    assertThat(uow.events).isEqualTo(expectedUow.events)
    assertThat(uow.version).isEqualTo(expectedUow.version)

  }

  // end::create_customer_test[]

  @Test
  @Throws(IOException::class)
  fun createCustomerIdempotency() {

    val customerId = CustomerId(UUID.randomUUID().toString())
    val createCustomerCmd = CreateCustomer(UUID.randomUUID(), customerId, "customer test")
    val expectedEvent = CustomerCreated(createCustomerCmd._targetId, "customer test")
    val expectedUow = EntityUnitOfWork(UUID.randomUUID(), createCustomerCmd,
            Version(1), listOf<DomainEvent>(expectedEvent))

    val json = mapper.writerFor(CreateCustomer::class.java).writeValueAsString(createCustomerCmd)

    log.info("command=\n" + json)

    val postCmdResponse = given().contentType(JSON).body(json)
            .`when`().post("/" + aggregateId(Customer::class.java) + "/commands")
            .then().extract().response()

    assertThat(postCmdResponse.statusCode()).isEqualTo(201)
    assertThat(postCmdResponse.header(LOCATION_HEADER))
            .isEqualTo(RestAssured.baseURI + ":" + RestAssured.port + "/"
                    + aggregateId(Customer::class.java) + "/commands/" + createCustomerCmd.commandId.toString())

    val getUowResponse = given().contentType(JSON).body(json)
            .`when`().get(postCmdResponse.header(LOCATION_HEADER))
            .then().statusCode(200).contentType(ContentType.JSON)
            .extract().response()

//    log.info("response -------------------")
//    log.info(getUowResponse.asString())

    val uow = mapper.readValue(getUowResponse.asString(), EntityUnitOfWork::class.java)

    assertThat(uow.targetId()).isEqualTo(expectedUow.targetId())
    assertThat(uow.command).isEqualTo(expectedUow.command)
    assertThat(uow.events).isEqualTo(expectedUow.events)
    assertThat(uow.version).isEqualTo(expectedUow.version)

    // now lets post it again

    val postCmdResponse2 = given().contentType(JSON).body(json).
            `when`().post("/" + aggregateId(Customer::class.java) + "/commands")
            .then().extract().response()

    assertThat(postCmdResponse2.statusCode()).isEqualTo(201)
    assertThat(postCmdResponse2.header(LOCATION_HEADER))
            .isEqualTo(RestAssured.baseURI + ":" + RestAssured.port + "/"
                    + aggregateId(Customer::class.java) + "/commands/" + createCustomerCmd.commandId.toString())

    val getUowResponse2 = given().contentType(JSON).body(json)
            .`when`().get(postCmdResponse2.header(LOCATION_HEADER))
            .then().statusCode(200).contentType(ContentType.JSON)
            .extract().response().asString()

    val uow2 = mapper.readValue(getUowResponse2, EntityUnitOfWork::class.java)

//    log.info("response 2-------------------")
//    log.info(getUowResponse2.toString())

    assertThat(uow).isEqualTo(uow2)
  }

  @Test
  @Throws(IOException::class)
  fun unknownCustomer() {

    val customerId = CustomerId(UUID.randomUUID().toString())
    val activateCustomer = ActivateCustomer(UUID.randomUUID(), customerId, "customer test")
    val json = mapper.writerFor(ActivateCustomer::class.java).writeValueAsString(activateCustomer)

    given().contentType(JSON).body(json)
            .`when`().put("/" + aggregateId(Customer::class.java) + "/commands")
            .then().statusCode(404)

  }

  companion object {

    val log = LoggerFactory.getLogger(Example1AcceptanceIT::class.java.simpleName)
  }

}